### ISSUE TITLE
Very slow nested queries compared to Prisma1

### ISSUE THREAD

**Issue by Frozen666** [prisma/prisma#2481]
## Bug description

Deeply nested queries are very slow compared to Prisma1. The example shown below was taking 500-800ms on Prisma1, on Prisma2 it takes 1.5-3s.

## How to reproduce

1. The schema was sent to schemas@prisma.io just now. We are using a POSTGRESQL DB, and running `2.0.0-beta.5`.
2. The resolver is simple: 
```js
export const initiativeConfigsQuery = queryField('initiativeConfigs', {
  type: 'InitiativeConfig',
  list: true,
  async resolve(parent, args, ctx) {
    const id = getUserId(ctx)
    return ctx.prisma.initiativeConfig.findMany({
      where: { project: { users: { some: { id } } } },
    })
  },
})
```
3. The query not so much:
<details>
<summary>client query</summary>

    {
      initiativeConfigs {
        id
        initiativeLayout {
          id
          groups {
            id
            labelKey
            blocks {
              id
              labelKey
              classes
              visibilityRule {
                property {
                  id
                }
                hierarchyElement {
                  id
                }
                option {
                  id
                }
                optionValue
                dateMin
                dateMax
                regex
                boolean
                numberMin
                numberMax
                use
                check
              }
              chart {
                id
              }
              blocks {
                id
                labelKey
                classes
                visibilityRule {
                  property {
                    id
                  }
                  hierarchyElement {
                    id
                  }
                  option {
                    id
                  }
                  optionValue
                  dateMin
                  dateMax
                  regex
                  boolean
                  numberMin
                  numberMax
                  use
                  check
                }
                chart {
                  id
                }
                blocks {
                  id
                  labelKey
                  classes
                  visibilityRule {
                    property {
                      id
                    }
                    hierarchyElement {
                      id
                    }
                    option {
                      id
                    }
                    optionValue
                    dateMin
                    dateMax
                    regex
                    boolean
                    numberMin
                    numberMax
                    use
                    check
                  }
                  chart {
                    id
                  }
                  blocks {
                    id
                    labelKey
                    classes
                    visibilityRule {
                      property {
                        id
                      }
                      hierarchyElement {
                        id
                      }
                      option {
                        id
                      }
                      optionValue
                      dateMin
                      dateMax
                      regex
                      boolean
                      numberMin
                      numberMax
                      use
                      check
                    }
                    chart {
                      id
                    }
                    blocks {
                      id
                      labelKey
                      classes
                      visibilityRule {
                        property {
                          id
                        }
                        hierarchyElement {
                          id
                        }
                        option {
                          id
                        }
                        optionValue
                        dateMin
                        dateMax
                        regex
                        boolean
                        numberMin
                        numberMax
                        use
                        check
                      }
                      chart {
                        id
                      }
                      blocks {
                        id
                        labelKey
                        classes
                        visibilityRule {
                          property {
                            id
                          }
                          hierarchyElement {
                            id
                          }
                          option {
                            id
                          }
                          optionValue
                          dateMin
                          dateMax
                          regex
                          boolean
                          numberMin
                          numberMax
                          use
                          check
                        }
                        chart {
                          id
                        }
                        primitives {
                          componentType
                          layoutShape {
                            defaultClass
                            defaultColor
                            rules {
                              order
                              class
                              color
                              rule {
                                property {
                                  id
                                }
                                hierarchyElement {
                                  id
                                }
                                option {
                                  id
                                }
                                optionValue
                                dateMin
                                dateMax
                                regex
                                boolean
                                numberMin
                                numberMax
                                use
                                check
                              }
                            }
                          }
                          selectType
                          imageUrl
                          textPrimitiveType
                          displayType
                          lockingRule {
                            property {
                              id
                            }
                            hierarchyElement {
                              id
                            }
                            option {
                              id
                            }
                            optionValue
                            dateMin
                            dateMax
                            regex
                            boolean
                            numberMin
                            numberMax
                            use
                            check
                          }
                          visibilityRule {
                            property {
                              id
                            }
                            hierarchyElement {
                              id
                            }
                            option {
                              id
                            }
                            optionValue
                            dateMin
                            dateMax
                            regex
                            boolean
                            numberMin
                            numberMax
                            use
                            check
                          }
                          labelKey
                          property {
                            pRules {
                              labelKey
                              hierarchyElement {
                                id
                              }
                              option {
                                id
                              }
                              optionValue
                              dateMin
                              dateMax
                              regex
                              boolean
                              numberMin
                              numberMax
                              use
                              check
                            }
                            formulaOnUpdate
                            hierarchy {
                              id
                            }
                            optionGroup {
                              id
                            }
                            defaultValue {
                              booleanValue
                              dateValue
                              numberValue
                              textValue
                              optionValue {
                                id
                              }
                            }
                            pType
                            description
                            labelKey
                            id
                          }
                          componentType
                          id
                        }
                      }
                      primitives {
                        componentType
                        layoutShape {
                          defaultClass
                          defaultColor
                          rules {
                            order
                            class
                            color
                            rule {
                              property {
                                id
                              }
                              hierarchyElement {
                                id
                              }
                              option {
                                id
                              }
                              optionValue
                              dateMin
                              dateMax
                              regex
                              boolean
                              numberMin
                              numberMax
                              use
                              check
                            }
                          }
                        }
                        selectType
                        imageUrl
                        textPrimitiveType
                        displayType
                        lockingRule {
                          property {
                            id
                          }
                          hierarchyElement {
                            id
                          }
                          option {
                            id
                          }
                          optionValue
                          dateMin
                          dateMax
                          regex
                          boolean
                          numberMin
                          numberMax
                          use
                          check
                        }
                        visibilityRule {
                          property {
                            id
                          }
                          hierarchyElement {
                            id
                          }
                          option {
                            id
                          }
                          optionValue
                          dateMin
                          dateMax
                          regex
                          boolean
                          numberMin
                          numberMax
                          use
                          check
                        }
                        labelKey
                        property {
                          pRules {
                            labelKey
                            hierarchyElement {
                              id
                            }
                            option {
                              id
                            }
                            optionValue
                            dateMin
                            dateMax
                            regex
                            boolean
                            numberMin
                            numberMax
                            use
                            check
                          }
                          formulaOnUpdate
                          hierarchy {
                            id
                          }
                          optionGroup {
                            id
                          }
                          defaultValue {
                            booleanValue
                            dateValue
                            numberValue
                            textValue
                            optionValue {
                              id
                            }
                          }
                          pType
                          description
                          labelKey
                          id
                        }
                        componentType
                        id
                      }
                    }
                    primitives {
                      componentType
                      layoutShape {
                        defaultClass
                        defaultColor
                        rules {
                          order
                          class
                          color
                          rule {
                            property {
                              id
                            }
                            hierarchyElement {
                              id
                            }
                            option {
                              id
                            }
                            optionValue
                            dateMin
                            dateMax
                            regex
                            boolean
                            numberMin
                            numberMax
                            use
                            check
                          }
                        }
                      }
                      selectType
                      imageUrl
                      textPrimitiveType
                      displayType
                      lockingRule {
                        property {
                          id
                        }
                        hierarchyElement {
                          id
                        }
                        option {
                          id
                        }
                        optionValue
                        dateMin
                        dateMax
                        regex
                        boolean
                        numberMin
                        numberMax
                        use
                        check
                      }
                      visibilityRule {
                        property {
                          id
                        }
                        hierarchyElement {
                          id
                        }
                        option {
                          id
                        }
                        optionValue
                        dateMin
                        dateMax
                        regex
                        boolean
                        numberMin
                        numberMax
                        use
                        check
                      }
                      labelKey
                      property {
                        pRules {
                          labelKey
                          hierarchyElement {
                            id
                          }
                          option {
                            id
                          }
                          optionValue
                          dateMin
                          dateMax
                          regex
                          boolean
                          numberMin
                          numberMax
                          use
                          check
                        }
                        formulaOnUpdate
                        hierarchy {
                          id
                        }
                        optionGroup {
                          id
                        }
                        defaultValue {
                          booleanValue
                          dateValue
                          numberValue
                          textValue
                          optionValue {
                            id
                          }
                        }
                        pType
                        description
                        labelKey
                        id
                      }
                      componentType
                      id
                    }
                  }
                  primitives {
                    componentType
                    layoutShape {
                      defaultClass
                      defaultColor
                      rules {
                        order
                        class
                        color
                        rule {
                          property {
                            id
                          }
                          hierarchyElement {
                            id
                          }
                          option {
                            id
                          }
                          optionValue
                          dateMin
                          dateMax
                          regex
                          boolean
                          numberMin
                          numberMax
                          use
                          check
                        }
                      }
                    }
                    selectType
                    imageUrl
                    textPrimitiveType
                    displayType
                    lockingRule {
                      property {
                        id
                      }
                      hierarchyElement {
                        id
                      }
                      option {
                        id
                      }
                      optionValue
                      dateMin
                      dateMax
                      regex
                      boolean
                      numberMin
                      numberMax
                      use
                      check
                    }
                    visibilityRule {
                      property {
                        id
                      }
                      hierarchyElement {
                        id
                      }
                      option {
                        id
                      }
                      optionValue
                      dateMin
                      dateMax
                      regex
                      boolean
                      numberMin
                      numberMax
                      use
                      check
                    }
                    labelKey
                    property {
                      pRules {
                        labelKey
                        hierarchyElement {
                          id
                        }
                        option {
                          id
                        }
                        optionValue
                        dateMin
                        dateMax
                        regex
                        boolean
                        numberMin
                        numberMax
                        use
                        check
                      }
                      formulaOnUpdate
                      hierarchy {
                        id
                      }
                      optionGroup {
                        id
                      }
                      defaultValue {
                        booleanValue
                        dateValue
                        numberValue
                        textValue
                        optionValue {
                          id
                        }
                      }
                      pType
                      description
                      labelKey
                      id
                    }
                    componentType
                    id
                  }
                }
                primitives {
                  componentType
                  layoutShape {
                    defaultClass
                    defaultColor
                    rules {
                      order
                      class
                      color
                      rule {
                        property {
                          id
                        }
                        hierarchyElement {
                          id
                        }
                        option {
                          id
                        }
                        optionValue
                        dateMin
                        dateMax
                        regex
                        boolean
                        numberMin
                        numberMax
                        use
                        check
                      }
                    }
                  }
                  selectType
                  imageUrl
                  textPrimitiveType
                  displayType
                  lockingRule {
                    property {
                      id
                    }
                    hierarchyElement {
                      id
                    }
                    option {
                      id
                    }
                    optionValue
                    dateMin
                    dateMax
                    regex
                    boolean
                    numberMin
                    numberMax
                    use
                    check
                  }
                  visibilityRule {
                    property {
                      id
                    }
                    hierarchyElement {
                      id
                    }
                    option {
                      id
                    }
                    optionValue
                    dateMin
                    dateMax
                    regex
                    boolean
                    numberMin
                    numberMax
                    use
                    check
                  }
                  labelKey
                  property {
                    pRules {
                      labelKey
                      hierarchyElement {
                        id
                      }
                      option {
                        id
                      }
                      optionValue
                      dateMin
                      dateMax
                      regex
                      boolean
                      numberMin
                      numberMax
                      use
                      check
                    }
                    formulaOnUpdate
                    hierarchy {
                      id
                    }
                    optionGroup {
                      id
                    }
                    defaultValue {
                      booleanValue
                      dateValue
                      numberValue
                      textValue
                      optionValue {
                        id
                      }
                    }
                    pType
                    description
                    labelKey
                    id
                  }
                  componentType
                  id
                }
              }
              primitives {
                componentType
                layoutShape {
                  defaultClass
                  defaultColor
                  rules {
                    order
                    class
                    color
                    rule {
                      property {
                        id
                      }
                      hierarchyElement {
                        id
                      }
                      option {
                        id
                      }
                      optionValue
                      dateMin
                      dateMax
                      regex
                      boolean
                      numberMin
                      numberMax
                      use
                      check
                    }
                  }
                }
                selectType
                imageUrl
                textPrimitiveType
                displayType
                lockingRule {
                  property {
                    id
                  }
                  hierarchyElement {
                    id
                  }
                  option {
                    id
                  }
                  optionValue
                  dateMin
                  dateMax
                  regex
                  boolean
                  numberMin
                  numberMax
                  use
                  check
                }
                visibilityRule {
                  property {
                    id
                  }
                  hierarchyElement {
                    id
                  }
                  option {
                    id
                  }
                  optionValue
                  dateMin
                  dateMax
                  regex
                  boolean
                  numberMin
                  numberMax
                  use
                  check
                }
                labelKey
                property {
                  pRules {
                    labelKey
                    hierarchyElement {
                      id
                    }
                    option {
                      id
                    }
                    optionValue
                    dateMin
                    dateMax
                    regex
                    boolean
                    numberMin
                    numberMax
                    use
                    check
                  }
                  formulaOnUpdate
                  hierarchy {
                    id
                  }
                  optionGroup {
                    id
                  }
                  defaultValue {
                    booleanValue
                    dateValue
                    numberValue
                    textValue
                    optionValue {
                      id
                    }
                  }
                  pType
                  description
                  labelKey
                  id
                }
                componentType
                id
              }
            }
            indicatorTable {
              defaultColumnWidth
              defaultPeriodType {
                id
              }
              property {
                hierarchy {
                  id
                  labelKey
                  root {
                    id
                  }
                  elements {
                    id
                    labelKey
                    children {
                      id
                    }
                    order
                    indicator {
                      id
                      labelKey
                      unitLabelKey
                      periodType {
                        id
                      }
                      hierarchyElement {
                        id
                      }
                      isSummable
                      allowAggregatedInput
                    }
                  }
                }
              }
            }
            childInitiativesProperty {
              childConfig {
                id
              }
              id
            }
            hideTab
            classes
            visibilityRule {
              property {
                id
              }
              hierarchyElement {
                id
              }
              option {
                id
              }
              optionValue
              dateMin
              dateMax
              regex
              boolean
              numberMin
              numberMax
              use
              check
            }
          }
          mainLayoutShape {
            defaultClass
            defaultColor
            rules {
              order
              class
              color
              rule {
                property {
                  id
                }
                hierarchyElement {
                  id
                }
                option {
                  id
                }
                optionValue
                dateMin
                dateMax
                regex
                boolean
                numberMin
                numberMax
                use
                check
              }
            }
          }
          initiativeName {
            componentType
            layoutShape {
              defaultClass
              defaultColor
              rules {
                order
                class
                color
                rule {
                  property {
                    id
                  }
                  hierarchyElement {
                    id
                  }
                  option {
                    id
                  }
                  optionValue
                  dateMin
                  dateMax
                  regex
                  boolean
                  numberMin
                  numberMax
                  use
                  check
                }
              }
            }
            selectType
            imageUrl
            textPrimitiveType
            displayType
            lockingRule {
              property {
                id
              }
              hierarchyElement {
                id
              }
              option {
                id
              }
              optionValue
              dateMin
              dateMax
              regex
              boolean
              numberMin
              numberMax
              use
              check
            }
            visibilityRule {
              property {
                id
              }
              hierarchyElement {
                id
              }
              option {
                id
              }
              optionValue
              dateMin
              dateMax
              regex
              boolean
              numberMin
              numberMax
              use
              check
            }
            labelKey
            property {
              pRules {
                labelKey
                hierarchyElement {
                  id
                }
                option {
                  id
                }
                optionValue
                dateMin
                dateMax
                regex
                boolean
                numberMin
                numberMax
                use
                check
              }
              formulaOnUpdate
              hierarchy {
                id
              }
              optionGroup {
                id
              }
              defaultValue {
                booleanValue
                dateValue
                numberValue
                textValue
                optionValue {
                  id
                }
              }
              pType
              description
              labelKey
              id
            }
            componentType
            id
          }
        }
        listLayout {
          id
          columns {
            id
            labelKey
            order
            defaultWidth
            defaultVisibility
            cType
            property1 {
              id
              optionGroup {
                id
              }
              hierarchy {
                id
              }
            }
          }
        }
        preview {
          title {
            id
          }
          description {
            id
          }
          layoutShape {
            defaultClass
            defaultColor
            rules {
              order
              class
              color
              rule {
                property {
                  id
                }
                hierarchyElement {
                  id
                }
                option {
                  id
                }
                optionValue
                dateMin
                dateMax
                regex
                boolean
                numberMin
                numberMax
                use
                check
              }
            }
          }
        }
        entityLabel {
          key
        }
        tileScreen {
          title
          rows
          columns
          config {
            preview {
              title {
                id
              }
              description {
                id
              }
              layoutShape {
                defaultClass
                defaultColor
                rules {
                  order
                  class
                  color
                  rule {
                    property {
                      id
                    }
                    hierarchyElement {
                      id
                    }
                    option {
                      id
                    }
                    optionValue
                    dateMin
                    dateMax
                    regex
                    boolean
                    numberMin
                    numberMax
                    use
                    check
                  }
                }
              }
            }
          }
          tiles {
            entity {
              id
            }
            area
          }
        }
        kanbanProperties {
          id
          pType
          labelKey
          optionGroup {
            id
          }
        }
      }
    }
</details>

## Expected behavior

The query duration should be shorter.

## Environment & setup

- OS: Windows 10
- Database: Postgres
- Prisma: `2.0.0-beta.5`
- Node.js: 12.13.1

**Comment by janpio** [comment]
Can you please set `DEBUG=*` and post the full log output you get when running this? Thanks.

**Comment by Frozen666** [comment]
Sorry, that output is too large, I uploaded it to Slack instead.

**Comment by pantharshit00** [comment]
@Frozen666 Can you link the slack message 🙏

**Comment by nikolasburk** [comment]
The output is here: https://prisma.slack.com/archives/G014DSFGH8Q/p1589551280010900

**Comment by pantharshit00** [comment]
Hi,

Sorry for late reply here. The nexus model took some time for setup. 

In my testing I was unable to reproduce the slowness:
![image](https://user-images.githubusercontent.com/22195362/83683728-7a963b00-a603-11ea-9c62-6758f6e20192.png)

I am using the same version as you for testing and I also tried the latest alpha and performance is the same.


Now only thing I don't have is the amount of data that you have in the database. Can you please share that? Or least please tell us how many rows are there in each model or maybe paste the result of the query after it is finished :pray:

**Comment by peteralbert** [comment]
@pantharshit00 Thanks! Is the above result using the large 1000 line query mentioned in the above issue? If this is the case, we could provide you with a Postgres DB dump - it does not have much data, so should only be a few MB...

**Comment by pantharshit00** [comment]
Yes, I am using the exact same query

**Comment by pantharshit00** [comment]
With the data I am getting the response under 300ms:
![image](https://user-images.githubusercontent.com/22195362/85198254-c8e65200-b304-11ea-9db0-3fbb44e29df7.png)

The first request takes around 700ms when connection to db is established and engines are spawned but after that I am setting a consistent 300ms response time. 

The postgres instance I am using is on the same machine and my machine has `Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz` along with 8 GB of RAM.

I used `2.0.1` for these test maybe upgrade the prisma and nexus version that you are using. I also tested it out with `2.0.0-beta.5` and result was the same.

Also, when I build the typescript code and run server with `NODE_ENV=production` the response is under 200ms:
![image](https://user-images.githubusercontent.com/22195362/85198753-c7b72400-b308-11ea-9801-a26a0e6f1052.png)

**Comment by pantharshit00** [comment]
Code for my graphql server:
```ts
import { nexusPrismaPlugin } from "nexus-prisma";
import { GraphQLServer } from "graphql-yoga";
import { PrismaClient } from "@prisma/client";
import * as path from "path";
import { makeSchema, queryField } from "@nexus/schema";
import * as types from "./types";

const prisma = new PrismaClient({
  // log: ["query"],
});

const schema = makeSchema({
  types: [
    types,
    queryField("initiativeConfigs", {
      type: "InitiativeConfig",
      list: true,
      async resolve(_parent, _args, ctx) {
        // const id = getUserId(ctx);
        return ctx.prisma.initiativeConfig.findMany({
          where: {
            project: { users: { some: { id: "cka3wzw9w0230w4ub4a7me175" } } },
          },
          include: {
            initiativeLayout: {
              include: {
                groups: {
                  orderBy: { order: "asc" },
                  include: {
                    blocks: {
                      orderBy: { order: "asc" },
                      include: {
                        primitives: {
                          orderBy: { order: "asc" },
                        },
                        blocks: {
                          orderBy: { order: "asc" },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        });
      },
    }),
  ],
  typegenAutoConfig: {
    contextType: "{ prisma: PrismaClient.PrismaClient }",
    sources: [{ source: ".prisma/client", alias: "PrismaClient" }],
  },
  outputs: {
    typegen: path.join(
      __dirname,
      "node_modules/@types/nexus-typegen/index.d.ts"
    ),
  },
  plugins: [nexusPrismaPlugin()],
});

const server = new GraphQLServer({
  context: () => ({ prisma }),
  schema,
});

server.start(() => {
  console.log(`🚀 GraphQL service ready at http://localhost:4000`);
});
```

**Comment by carlosfiori** [comment]
+1

**Comment by janpio** [comment]
What do you refer to @carlosfiori? Do you also have slow perceived slow query? Then best open a new issue and provide the same information @pantharshit00 asked for above - then we are happy to try to reproduce this as well.

**Comment by carlosfiori** [comment]
@janpio Hi,
That is exactly my problem, I have a graphql query that execute 11 sql query. Running that 11 queries on database took approx 1.5 seconds, but with prisma is taking 5 seconds.
Actually even a simple query with one field is take more than 1 second, and on database take less than 200ms

**Comment by gabrielcarneiro97** [comment]
same problem here, with Prisma a nested query is taking up to 8 seconds, but with manual query builder it takes something around 1.5 sec

**Comment by janpio** [comment]
@gabrielcarneiro97 :

As mentioned before:

> Then best open a new issue and provide the same information @pantharshit00 asked for above - then we are happy to try to reproduce this as well.

**Comment by SelfDevTV** [comment]
Hi,

has there been any updates on this one?

I also opened a discussion thread where my api is very slow (i use prisma to fetch data with some `includes`)
It's this one #4878 maybe that helps it has much informations in it.

**Comment by pantharshit00** [comment]
#4878 is resolved. 

Closing this issue due to inactivity.
